### DISCUSSION TITLE
myproject.page - A project showcase website

### DISCUSSION THREAD

**Discussion by championswimmer** [scaleracademy/open-source-projects#2] [Ideas]
## Summary

[myproject.page](http://myproject.page) is a website for people to showcase their projects and discover projects by peers. Unlike a website like Product Hunt for example, the focus is not on how unique or how viable (business wise) the product is, but on how well the project is made as far as technicalities and design goes. 

A growing segment of the developer/designer/product market is looking for people with past projects having been made. A portfolio site where projects can be showcased, and also where great projects can be discovered can help foster this community. 

### Portfolio Building

The first important part of this portal is to signup, create accounts and add projects. We will use 'login with Scaler' to hasten the signup process. 

Everyone gets a [myproject.page/@username](http://myproject.page/@username)  format profile landing page where there is basic information about them, plus a list of their projects. 

To add projects we need a high entry barrier, i.e. all details like title, description should have well defined min-max char count. Cover photo is important, options to add Youtube video, Github link etc should be available. 

### Browse Projects

Once projects exist on the website, it is important to be able to browse projects, upvote/downvote them and discuss about them 

We will take a imgur like approach where there will be pages like 

[myproject.page/new](http://myproject.page/new)  

[myproject.page/trending](http://myproject.page/trending) 

[myproject.page/popular](http://myproject.page/popular) 

Each project will be basically available in it's own landing page like [myproject.page/@username/proj-name-23b4j32](http://myproject.page/@username/proj-name-23b4j32) 

On the project page we can integrate Disqus so people can comment on the projects as well. (or our own comment system too) 

Browsing projects by hashtags is important as well. So that you can go to [myproject.page/#react](http://myproject.page/#react) for react projects. 

### Reporting Projects and Comments

- spam
- abuse
- duplicate

### Gamification & Badges

Examples 

- 1st project uploaded
- 5 projects uploaded
- Got 10 upvotes
- Got 100 upvotes

### Project Timeline

- update a project later after submissino
- show updates / additions to a project

### Emoji-based reactions for Projects 

Allow finding projects by college, so I can find projects of my college mates

**Comment by championswimmer** [comment]
# User Stories 

## Adding Content

- User Signup Flow
    - Login/Signup Page ( ignore for now, and use login with CB )
    - User Profile Edit
- Add Project Form
    - Title*, Description*, Summary, Photos*, Demo Video

## Browsing Content

- Home Page → Top Level Listings
    - New
    - Trending
    - Popular
- Single Project View (when we open one particular project)
    - Comments section - powered by Disqus
- User Profile Page

**Reply by championswimmer** [reply]
## Entities 

### 1. `users` 

| column          | definition                 |
|-----------------|----------------------------|
| id              | bigint (auto increment) PK |
| username        | string (30) - unique       |
| email           | string (email)             |
| page            | nth page eg: 5             |
| github_username | string                     |
| links           | array<string(url)>         |
| bio             | string (40 chars)          |
| avatar          | string (image url)         |
| follower_count  | bigint                     |
| followee_count  | bigint                     |

### 2. `projects` 

| column            | definition                   |
|-------------------|------------------------------|
| id                | bigint (auto increment) PK   |
| creator_id        | bigint (foreign key user.id) |
| slug              | string (50) - unique         |
| title             | string (100)                 |
| short_description | string (240)                 |
| long_description  | string (4000) - markdown     |
| images            | array<string(url)>           |
| source_links      | array<string (url)>          |
| website           | string (url)                 |
| upvote_count      | bigint                       |
| comment_count     | bigint                       |
| hashtags | array<string>| 

### 3. `project_upvotes` 
| column            | definition                      |
|-------------------|---------------------------------|
| id                | bigint (auto increment) PK      |
| project_id        | bigint (foreign key project.id) |
| upvoter_id        | bigint (foreign key user.id)    |

### 4. `comments` 
| column        | definition                      |
|---------------|---------------------------------|
| id            | bigint (auto increment) PK      |
| project_id    | bigint (foreign key project.id) |
| commenter_id  | bigint (foreign key user.id)    |
| title         | string (100)                    |
| body          | string (500)                    |
| reply_to      | bigint (foreign key comment.id) |
| upvote_count  | bigint                          |

### 3. `comment_upvotes` 
| column            | definition                      |
|-------------------|---------------------------------|
| id                | bigint (auto increment) PK      |
| comment_id        | bigint (foreign key comment.id) |
| upvoter_id        | bigint (foreign key user.id)    |

### 4. `hashtags` 

| column            | definition                      |
|-------------------|---------------------------------| 
| id                | bigint (auto increment) PK      |
| name    | string (unique) | 
| type    |    enum ( language, framework, topic) | 

example: `#javascript` is language hashtag 
example: `#react` is framework hashtag 
example: `#fintech` is topic hashtag 

### 5. project_hashtags 
| column            | definition                      |
|-------------------|---------------------------------| 
| hashtag_id                | bigint (foreign key hashtag.id)    |
| project_id                | bigint (foreign key project.id)    |

**Reply by championswimmer** [reply]
# API Endpoints 

## `projects` 

### `GET /projects` 📃


| query         | definition                           |
|---------------|--------------------------------------|
| sort          | new \| popular \| trending (default)           |
| tags          | show projects of a paritcular hashtag |

### `POST /projects` 🔒
Creates a new project 

### `GET /users/@{username}/projects` 
show projects of a given username 

### `GET /projects/{project slug}` 
get details of 1 particular project 

## `comments` 

### `GET /projects/{slug}/comments` 📃 
show all comments under a project 

### `POST /projects/{slug}/comments` 🔒
add a comment under a given project

## `users` 

### `GET /users`  📃 
show all users 

### `POST /users` 
create a new user 

### `GET /users/@{username}` 
get details of a particular user 

### `PATCH /users/@{username}`  🔒
update details of an user 

## `upvotes` 

### `PUT /projects/{slug}/upvote`  🔒

### `DELETE /projects/{slug}/upvote`  🔒

## `hashtags` 

### `GET /hashtags` 📃
get all hashtags 

### `POST /hashtags` 🔏
create a new hashtag

**Comment by shagun-tech** [comment]
URL not working

**Reply by championswimmer** [reply]
@shagun-tech we have the domain, the project is not yet complete or deployed on it though. 

Would like to see more discussion around pros/cons of the idea and suggestions on features before we start working on this project.

**Comment by Chirag482** [comment]
Entry barrier can aslo include an optional option to guide the project through a tech blog with URL myproject.page/@username/proj-name-23b4j32-blog and such guided projects can be searched by using  myproject.page/#guided-projects.

**Reply by championswimmer** [reply]
You mean the blog post should be a walkthrough of the project ?

**Reply by Chirag482** [reply]
The blog post must explain the step by step procedure of making the project.
example -> https://reactjs.org/tutorial/tutorial.html#lifting-state-up  guiding to build a Tic Tac Toe.

**Comment by sPredictorX1708** [comment]
This is already done by devfolio right!!

**Reply by dewanshrawat15** [reply]
@championswimmer each DevFolio user is assigned a url like devfolio.co/@dewanshrawat15 where you can view hackathons the user participated in or the projects the user built. Each project is very well documented and gets a seperate URL to showcase.

**Reply by championswimmer** [reply]
That is only for hackathons, not for in-general side projects you are making, outside of hackathons.

**Reply by sPredictorX1708** [reply]
https://guide.devfolio.co/hackers/devfolio-profile/projects-tab

Projects tab I guess is for showcasing !

**Comment by shubhamsingla807** [comment]
Great Idea, initially in my 2nd year of grad I also want to build something like that and designed some screens for that. (But due to time constraints I leave that project). Apart from only showcasing projects, I thought of adding few more things to it (Also it's for both developers and designers) - 

1. Discussions Page - Everyone can discuss challenges/anything they want to discuss in a forum-like thing. Community-based learnings are always great.

2. Mentoring - A lot of people can't able to make it big because they don't know what should they do, they are unclear of goals, so I thought industry mentors could spare some time and mentor them (Same like https://www.adplist.org/mentors, it's for only designers though). But here as I told earlier it's a platform for both dev and design folks.

3. Projects showcase is also there, but what is we can make a feature in which they can share their profiles easily to anybody (It becomes a sort of portfolio for them). (There are a lot more things that can go in this, but maybe if you're building this project then I can contribute and discuss more).

4. Challenges - As I told it's a platform for both dev and designers, maybe designers can challenge other people to build this and vice versa.

I know there are a lot of things but this could be a great project, I'm attaching a few screenshots too that I designed) -

![image](https://user-images.githubusercontent.com/43694495/120966804-5aeca200-c784-11eb-97f8-b3f466732607.png)

**Reply by championswimmer** [reply]
Nice, @shubhamsingla807 I am sure you can be one of the most important contributors of this project :)

**Comment by vighnesh192** [comment]
Glad to see somebody is still stuck to this one!!

Please check these out:- 

<p align="center">
  <img src="https://user-images.githubusercontent.com/64535868/122119881-5c366280-ce47-11eb-98a3-36aff655a4d7.png" width="45%">
&nbsp; &nbsp; &nbsp; &nbsp;
  <img src="https://user-images.githubusercontent.com/64535868/122119958-7708d700-ce47-11eb-89a5-6a3666b56011.png" width="45%">
<br>
<br>
<img src="https://user-images.githubusercontent.com/64535868/122119965-796b3100-ce47-11eb-9062-ad60f3371dd8.png" width="45%">
&nbsp; &nbsp; &nbsp; &nbsp;
<img src="https://user-images.githubusercontent.com/64535868/122119975-7c662180-ce47-11eb-88a3-ee7858b204ed.png" width="45%">
</p>

Client Repo:- https://github.com/vighnesh192/project-showcase-client
Server Repo:- https://github.com/vighnesh192/project_showcase_server

The problem is setting up the server is a bit time consuming and the contributor will have to go through the entire process even though they are here just for the front end. I'm trying to figure this out and I really want this to go live!

**Reply by championswimmer** [reply]
@vighnesh192 these look really good man! Possible for you to host this up on Heroku or something ? Would love to see it in action, and possible best if we can fork/pull these in, if you wish to bring this project in and be maintainer for it :) Since you have put in so much effort already into it, can pretty much start from there itself instead from a blank slate.

**Reply by vighnesh192** [reply]
Thank you! Sure, I'll host it on Heroku and post it here.
Actually I've created a separate org account for this...so I was gonna shift these repos there once it is ready (MVP at least)😅

**Reply by vighnesh192** [reply]
Deployed to Heroku:- 
https://projstemp.herokuapp.com/
